File: R/04_phase_maps.R
#' 04_phase_maps.R
#' Functions to generate phase transition maps from parameter sweeps.

library(dplyr)
library(ggplot2)
library(tidyr)

source("01_load_data.R")

#' Create Phase Map of Winning Policies
#'
#' @param df Dataframe from parameter sweeps containing multiple policies
#' @param x_var Character name of the parameter mapped to the x-axis
#' @param y_var Character name of the parameter mapped to the y-axis
#' @param z_var Variable defining success (defaults to "cumulative_reward")
#' @param grid_resolution Optional integer to bin continuous parameters into a grid
#' @return A ggplot object showing the policy that "wins" at each parameter coordinate.
plot_policy_phase_map <- function(df, x_var, y_var, z_var = "cumulative_reward", grid_resolution = NULL) {
  
  df_map <- df |>
    drop_na(all_of(c(x_var, y_var, z_var, "policy")))
  
  if (!is.null(grid_resolution)) {
    # Bin continuous parameters to create discrete tiles
    df_map <- df_map |>
      mutate(
        !!sym(x_var) := cut(!!sym(x_var), breaks = grid_resolution, labels = FALSE),
        !!sym(y_var) := cut(!!sym(y_var), breaks = grid_resolution, labels = FALSE)
      )
  }
  
  # Find the winning policy for each (x, y) coordinate
  winners <- df_map |>
    group_by(across(all_of(c(x_var, y_var, "policy")))) |>
    summarize(mean_score = mean(!!sym(z_var), na.rm = TRUE), .groups = "drop") |>
    group_by(across(all_of(c(x_var, y_var)))) |>
    slice_max(order_by = mean_score, n = 1, with_ties = FALSE) |>
    ungroup()
  
  p <- ggplot(winners, aes(x = .data[[x_var]], y = .data[[y_var]], fill = policy)) +
    geom_tile(color = "white") +
    theme_minimal() +
    labs(
      title = paste("Phase Map: Dominant Policy by", x_var, "and", y_var),
      x = x_var,
      y = y_var,
      fill = "Winning Policy"
    ) +
    scale_fill_viridis_d(option = "plasma")
  
  return(p)
}

#' Create Heatmap of Continuous Variable
#'
#' @param df Dataframe from load_parquet_results
#' @param x_var X-axis parameter
#' @param y_var Y-axis parameter
#' @param z_var Value to show in the heatmap (e.g., population_gini)
#' @return A ggplot heatmap
plot_outcome_heatmap <- function(df, x_var, y_var, z_var = "population_gini") {
  
  # Average z_var across seeds for each x, y combination
  summary_df <- df |>
    group_by(across(all_of(c(x_var, y_var)))) |>
    summarize(mean_val = mean(!!sym(z_var), na.rm = TRUE), .groups = "drop")
  
  p <- ggplot(summary_df, aes(x = .data[[x_var]], y = .data[[y_var]], fill = mean_val)) +
    geom_tile() +
    scale_fill_viridis_c() +
    theme_minimal() +
    labs(
      title = paste("Heatmap of", z_var),
      x = x_var,
      y = y_var,
      fill = paste("Mean", z_var)
    )
  
  return(p)
}

# Example Usage:
# df_sweep <- load_parquet_results("../results/sweeps/shock_intensity_vs_resource_abundance")
# p_phase <- plot_policy_phase_map(df_sweep, "shock_probability", "resource_rent_multiplier")
# ggsave("policy_phase_map.png", p_phase)
# p_heat <- plot_outcome_heatmap(df_sweep, "shock_probability", "resource_rent_multiplier", "population_gini")

File: python/04_phase_maps.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from typing import Optional

def plot_policy_phase_map(
    df: pd.DataFrame, 
    x_var: str, 
    y_var: str, 
    z_var: str = 'cumulative_reward', 
    grid_resolution: Optional[int] = None
) -> None:
    """Create a phase map showing the dominant policy across parameter sweeps."""
    df_map = df.dropna(subset=[x_var, y_var, z_var, 'policy']).copy()
    
    if grid_resolution is not None:
        # Bin continuous parameters
        df_map[x_var] = pd.cut(df_map[x_var], bins=grid_resolution, labels=False)
        df_map[y_var] = pd.cut(df_map[y_var], bins=grid_resolution, labels=False)
        
    # Group by parameters and policy to find average reward
    grouped = df_map.groupby([x_var, y_var, 'policy'], observed=True)[z_var].mean().reset_index()
    
    # Find winning policy for each coordinate
    idx = grouped.groupby([x_var, y_var])[z_var].idxmax()
    winners = grouped.loc[idx].copy()
    
    # Pivot for heatmap
    heatmap_data = winners.pivot(index=y_var, columns=x_var, values='policy')
    
    plt.figure(figsize=(10, 8))
    
    # Map policies to colors manually since seaborn heatmap doesn't handle categoricals easily
    policies = sorted(df_map['policy'].unique())
    n_colors = len(policies)
    cmap = sns.color_palette("plasma", n_colors)
    color_dict = {p: i for i, p in enumerate(policies)}
    
    # Convert policy names to integers
    num_data = heatmap_data.replace(color_dict)
    
    ax = sns.heatmap(
        num_data, 
        cmap=cmap, 
        cbar_kws={'ticks': range(n_colors)}
    )
    
    # Format colorbar labels
    colorbar = ax.collections[0].colorbar
    colorbar.set_ticklabels(policies)
    colorbar.set_label('Winning Policy')
    
    plt.title(f"Phase Map: Dominant Policy by {x_var} and {y_var}")
    ax.invert_yaxis()  # Standard orientation for y-axis
    
    return plt.gcf()

def plot_outcome_heatmap(df: pd.DataFrame, x_var: str, y_var: str, z_var: str = 'population_gini') -> None:
    """Create a heatmap of a continuous outcome variable across a parameter space."""
    summary_df = df.groupby([x_var, y_var], observed=True)[z_var].mean().reset_index()
    
    heatmap_data = summary_df.pivot(index=y_var, columns=x_var, values=z_var)
    
    plt.figure(figsize=(10, 8))
    ax = sns.heatmap(heatmap_data, cmap='viridis')
    ax.invert_yaxis()
    
    plt.title(f"Heatmap of {z_var}")
    plt.xlabel(x_var)
    plt.ylabel(y_var)
    
    return plt.gcf()
